src/polyswarm_api/types/artifact.py

```diff
@@ -240,6 +243,11 @@ def hash(self):
     def hash_type(self):
         return "sha256"
 
+    @property
```

Comment:
I would personally not term it as abuse. Almost all of the instances where it is used, it describes a property of the object, just one that is lazily calculated and/or fetched. I’m curious though your reasoning for why that would be “abusive”. I believe this is the intended usage of the property decorator? We could make them normal methods, but syntactically that seems silly to me.

Currently there is no distinction really between “high level” and “low level” types. The API types are what is returned from the API, with some convenience methods and properties to make it much simpler to use. If these convenience properties were returned in the actual API response it would be easier, but I’m very concerned about making interacting with these objects more difficult when accessing common information. I don’t want or expect people to dig three layers deep into nested objects or manually invoke API calls just to fetch more data about the object you already have.

Reply:
Many of these attributes are not trivial calculations over existing attributes of the class. When a developer sees an attribute, he expects that it will be already there, calculated and available. We are hiding a lot of logic, which sometimes make extra external calls, under the hood of a simple attribute access. This reduces the ability of the user to tune his code and know when what they are doing is costly or not.

This also creates the wrong impression that these features are available from the api, which is not the case. The user will use the client and assume that somehow this information is available from the endpoint, since it is readily available in the client just as other fields that they know exist in the api are available. This will create a gap of expectations of what is in the api and what is in the client (even I created expectations when I was reading the code, I was looking at some things being accessed in the objects and I was "wait, is this in the api?"). People should get used the how the resources we provide are designed and learn how to navigate them, and always make it clear the distinction between a helper method assembling stuff in a convenient way from the resource structure we actually have. If we think some of these things are that important, they should be added to the api rather than calculated client-side (I am not talking about nested objects here, these should be a helper method that is identified as such).

My reasoning behind abusive is that it does not allow me to get intuition on how that information is being generated, meaning that it is running expansive code to get it or fetching it from the api. It also does not give me intuition from how the api is designed, since what is api-sourced and what is not is not clear. Because of this, I might write bad code since I don't know what I am manipulating, something that looks innocent as an attribute lookup can become a huge bottleneck. I think the word "abusive" might not be the most adequate, I did not mean to be offensive. I would rather see these as sintax improvements that can give hints to the user of the inner workings without having to make him look at it.

Some examples of how I like to use properties are the following (and I give you this is highly personal):

```python
class Foo:
    def __init__(self, raw_object):
        self._remote_attr = None
        self.raw_object = None
        self._object = None
    
    @property
    def remote_attr(self):
        if self._remote_attr is None:
            self._remote_attr = some_remote_call()
        return self._remote_attr

    @property
    def object(self):
        if self._object is None:
            self._object = some_expansive_transformation(self.raw_object)
        return self._object
```

I focus on values that can be cached most of the time, and use it for lazy evaluation and set the result of this evaluation to a internal attribute. I also focus on transformations of existing data in the current object, meaning that this is just transforming data that is already relevant to the current instance.

The places where I see these improvement could be most beneficial are such as these:

```
    @property
    def scans(self):
        # do not report scans as they are running, only once window has closed
        return list(filter(None, [instance for instance in self.instances
                                  if instance.window_closed and not instance.failed]))

    @property
    def scan_permalink(self):
        if len(self.bounties) == 0:
            return None
        return self.instances[0].submission_uuid

    @property
    def bounties(self):
        return [instance.bounty for instance in self.instances if instance.bounty]

    @property
    def filenames(self):
        """ Unique filenames in all observed instances"""
        return list(OrderedSet([instance.name for instance in self.instances if instance.name]))

    @property
    def countries(self):
        return list(OrderedSet([instance.country for instance in self.instances if instance.country]))
```

These, for instance, make it look like the `Artifact` attribute have these attributes by default. When using such properties, I would imagine that the API somehow provides these to the client. It also makes a data transformation on data that belongs to another object. The user will not learn that `instance`s hold this information (not learning about our resource structure) and might think that this is readily available, writing code that relies heavily on this (and since this will hopefully be paginated, it will not be really efficient as is).

Sorry for the long comment, we can totally have a chat about this later, I think it might be more productive.